### DISCUSSION TITLE
TASMOTA Hand Wave detection

### DISCUSSION THREAD

**Discussion by sivabala21** [arendst/Tasmota#20441] [Q&A] [answered]
I am using IR sensor to detects the motion of handwave detection.
I connected it as switch and decoupled and setted it as toggle since its for handwave detection
Now i decoupled and wrote rule to when toogle happen turn on/off 
but here the problem
Actually i want this logic in my handwave detection
if one realy ON and one realy OFF when sensor trigger turn OFF the both relay
if both of the relay OFF then turn ON the both of them
if both of the relay ON then turn OFF the both of them

I generated rule using bard but its showing error and not working as expected

Rule1 ON Switch1#State=2 DO Backlog IF (POWER1#STATE=0) && (POWER2#STATE=0) DO POWER1 ON; POWER2 ON ELSE POWER1 OFF; POWER2 OFF ENDON; Delay 5; // Optional delay for hand wave detection Backlog POWER1 TOGGLE; POWER2 TOGGLE ENDON ENDON
Please Help me to fix the issue

**Comment by barbudor** [comment]
Then Bard hasn't read the doc correctly 🤣 
May be ChatGPT has ?

There are a lot of issues

## 1 -  `IF (POWER1#STATE=0) ` can't work
You (or Bard) is mistaking a trigger and a variable
`Power1#state` is a trigger that can **only** appear between `ON` and `DO`. It defines to which event triggers the rule
Only variables like `%var1%` , `%mem1%`, `%value%`  or other system variables such as `%time%`. 
But there is *no* variable like `%power1%` so you must do differently

## 2 - The over all structure is wrong
A rule must be `ON trigger DO something ENDON`
Your syntax is much away from that

Something like would be closer to what you are looking for (split on multiple lines for readability)
```
Rule1
  ON Power1#state DO var1 %value% ENDON
  ON Power2#state DO var2 %value% ENDON
  ON var1#state==1 DO IF (%var2%==0) ... ELSE  ... ENDON
  ...
```
I have'=n't understood your logic so I can't fill the `...` but you should get the idea
- Save the power value into a variable
- Trigger when 1 variable change and act based on the value of the other variable

I fear that rules won't be fast enough for what you are looking for
You may be interrested by Tasmota support for gesture sensors such as APDS9960 or PAJ7620
I have once tested APDS9960 and as far as I remember it was working fine, directly providing you with the gesture information

**Reply by sivabala21** [reply]
Thanks for Spending Time to Approch my Issue , Definitely I go through your recommandation .

**Reply by sivabala21** [reply]
Condition Checks:
- **Both relays off:** 
  - If both `var1` and `var2` are 0 (relays off), turn both relays on.

- **One relay on, one off:** 
  - If either `var1` or `var2` is 1 (one relay on), turn both relays off.

- **Both relays on:** 
  - If both `var1` and `var2` are 1 (both relays on), turn both relays off.


`var1` is POWER STATE (0/1) of relay 1
`var2` is POWER STATE (0/1) of relay 2


Consider "SENSOR" as a switch that operates using a toggle function - it turns on when it detects a hand and immediately turns off.
Can you give me the crafted rule ?

**Reply by barbudor** [reply]
Please ignore my previous answer
I mixup Switch and Relay/Power

but I think you too

**Reply by barbudor** [reply]
you need to think in terms of event
A rule is triggered on an event
In most cases when something change from one state to another state (or when a message is published)
so what you want to do is be triggered by Switch events, not power events, and act based on Power state

Note the usage of `Power0` which change all relay of the device at the same time.
Of course if you have more than 2 relay, you can't use that and need to use separated `Power1 OFF; Power2 OFF` 

Something like thebelow seems to match your description

```
Rule1 
  ON Power1#state DO var1 %value% ENDON
  ON Power2#state DO var2 %value% ENDON
  ON Switch1#state==0 DO IF (%var2%==0) THEN Power0 ON ENDON
  ON Switch2#state==0 DO IF (%var1%==0) THEN Power0 ON ENDON
  ON Switch1#state==1 DO Power0 OFF ENDON
  ON Switch2#state==1 DO Power0 OFF ENDON
```


Another option that came to my mind is that you could probably use your IR switch with a Rotary Encoder
Depending on the order the GPIO would be triggered, Tasmota will automatically increment or decrement the value and you can catch that in a rule

**Reply by sfromis** [reply]
Actually, you **do** have `%power1%` and so on variables for power channels. Simpler than going via rules to catch them.
https://tasmota.github.io/docs/Rules/#rule-variables

`%switch1%` and so on are also available, if convenient.

Still, using AI for Tasmota rules is best if you *want* to get bad advice.

**Reply by barbudor** [reply]
Ah yes.
I had that something in the corner of my mind but I checked the code, not the doc. I didn't look at the right place.

So the above could be simplified as:
```
Rule1
  ON Switch1#state==0 DO IF (%power2%==0) Power0 ON ENDON
  ON Switch2#state==0 DO IF (%power1%==0) Power0 ON ENDON
  ON Switch1#state==1 DO Power0 OFF ENDON
  ON Switch2#state==1 DO Power0 OFF ENDON
```

But still I strongly doubt that this is the best approach to the problem

*Edit as per sfromis comment below*

**Reply by sfromis** [reply]
Assuming a build with that feature, correct syntax would be:
`ON Switch1#state==0 DO IF (%power2%==0) Power0 ON ENDIF ENDON`

Whether it is a good approach is another question, of course.

**Reply by barbudor** [reply]
Apparently I'm tired 😄

**Comment by sivabala21** [comment, marked as answer]
**Thanks for the assistance! I was able to craft the following rule with the help of your suggestions and the documentation:**

```
Rule1 ON Switch1#state==2 DO
  IF ((%power2% == 0) AND (%power4% == 0))
    Backlog Power2 1; Power4 1
  ELSEIF ((%power2% == 1) AND (%power4% == 1))
    Backlog Power2 0; Power4 0
  ELSE
    Backlog Power2 0; Power4 0
  ENDIF
ENDON
```



**Important note for ESP8266 users:**

- This rule requires a custom Tasmota build with IF and expression features enabled. It will not work with the default ESP8266 build.

**Reply by Jason2866** [reply]
The tasmota-4M build (yes this is an esp8266 build) do have if and expression enabled.
